common/lib/plugins/failover2/failover2_plugin.ts

```diff
@@ -0,0 +1,502 @@
+        await this.failover(hostInfo);
+      } catch (error) {
+        if (error instanceof FailoverSuccessError) {
+          client = this.pluginService.getCurrentClient().targetClient;
+        } else {
+          throw error;
+        }
+      }
+    }
+
+    if (!client) {
+      // This should be unreachable, the above logic will either get a connection successfully or throw an exception.
+      throw new AwsWrapperError(Messages.get("Failover2.unableToConnect"));
+    }
+
+    if (isInitialConnection) {
+      await this.pluginService.refreshHostList(client);
+    }
+
+    return client;
+  }
+
+  override async execute<T>(methodName: string, methodFunc: () => Promise<T>): Promise<T> {
+    try {
+      // Verify there aren't any unexpected error emitted while the connection was idle.
+      if (this.pluginService.hasNetworkError()) {
+        // Throw the unexpected error directly to be handled.
+        throw this.pluginService.getUnexpectedError();
+      }
+
+      if (!this.enableFailoverSetting || this.canDirectExecute(methodName)) {
+        return await methodFunc();
+      }
+
+      if (this.isClosed) {
+        await this.invalidInvocationOnClosedConnection();
+      }
+
+      if (this.canUpdateTopology(methodName)) {
+        await this.updateTopology();
+      }
+
+      return await methodFunc();
+    } catch (e: any) {
+      logger.debug(Messages.get("Failover.detectedError", e.message));
+      if (this._lastError !== e && this.shouldErrorTriggerClientSwitch(e)) {
+        await this.invalidateCurrentClient();
+        const currentHostInfo = this.pluginService.getCurrentHostInfo();
+        if (currentHostInfo !== null) {
+          this.pluginService.setAvailability(currentHostInfo.allAliases ?? new Set(), HostAvailability.NOT_AVAILABLE);
+        }
+
+        this._lastError = e;
+        await this.pickNewConnection();
+      }
+
+      throw e;
+    }
+  }
+
+  async failover(failedHost: HostInfo) {
+    this.pluginService.setAvailability(failedHost.allAliases, HostAvailability.NOT_AVAILABLE); // checking if needed
+
+    if (this.failoverMode === FailoverMode.STRICT_WRITER) {
+      await this.failoverWriter();
+    } else {
+      await this.failoverReader();
+    }
+
+    if (this._isInTransaction || this.pluginService.isInTransaction()) {
+      // "Transaction resolution unknown. Please re-configure session state if required and try
+      // restarting transaction."
+      logger.debug(Messages.get("Failover.transactionResolutionUnknownError"));
+      throw new TransactionResolutionUnknownError(Messages.get("Failover.transactionResolutionUnknownError"));
+    } else {
+      // "The active SQL connection has changed due to a connection failure. Please re-configure
+      // session state if required."
+      throw new FailoverSuccessError(Messages.get("Failover.connectionChangedError"));
+    }
+  }
+
+  async failoverReader() {
+    const telemetryFactory = this.pluginService.getTelemetryFactory();
```

Comment:
Looks like Node's implementation is missing some logs, could we double check against JDBC